In [57]:
import nltk
from nltk import word_tokenize
from nltk.corpus import stopwords
import os
from sklearn.feature_extraction.text import TfidfVectorizer
import math


In [7]:
os.chdir('C:\BTS_Master\BTS_DataScienceFoundations/Session_7')

In [43]:
lyrics = []

# iterates between each file in the direct, if its a txt file reads it and appends to the lyrics list replacing '\n' with space
for file in os.listdir('C:\BTS_Master\BTS_DataScienceFoundations/Session_7'):
    if file.endswith('.txt'):
        with open(str(file)) as f:
            lyrics.append(f.read().replace('\n', ' '))


In [12]:
tokenize = lambda doc: doc.lower().split(" ")


In [17]:
tfidf = TfidfVectorizer(norm='l2', min_df=0, use_idf=True, smooth_idf=False, sublinear_tf=True, tokenizer=tokenize)

representation = tfidf.fit_transform(lyrics)

In [53]:
def cosine_similarity(vector1, vector2):
    dot_product = sum(p*q for p,q in zip(vector1, vector2))
    magnitude = math.sqrt(sum([val**2 for val in vector1])) * math.sqrt(sum([val**2 for val in vector2]))
    if not magnitude:
        return 0
    return dot_product/magnitude

In [56]:
songs_name = ['Through Her Eyes', 'The Spirit Carries On', "Beneath The Surface", "Pull Me Under", "Wither", 
              "Another Brick In The Wall (Part II)", "Hey You", "Comfortably Numb", 'Time', "Wish You Were Here" ]
for i in range(10):
    for j in range(i+1,10):
        similarity = cosine_similarity(representation.toarray()[i], representation.toarray()[j])
        print("The similarity between the song {} and the song {} is {}".format(songs_name[i], songs_name[j], similarity))

The similarity between the song Through Her Eyes and the song The Spirit Carries On is 0.1521289773032246
The similarity between the song Through Her Eyes and the song Beneath The Surface is 0.154101895634545
The similarity between the song Through Her Eyes and the song Pull Me Under is 0.15894772678581914
The similarity between the song Through Her Eyes and the song Wither is 0.1465945306005967
The similarity between the song Through Her Eyes and the song Another Brick In The Wall (Part II) is 0.06280735989387921
The similarity between the song Through Her Eyes and the song Hey You is 0.09928488990540278
The similarity between the song Through Her Eyes and the song Comfortably Numb is 0.19775882519287555
The similarity between the song Through Her Eyes and the song Time is 0.12905663990419475
The similarity between the song Through Her Eyes and the song Wish You Were Here is 0.08273714019971634
The similarity between the song The Spirit Carries On and the song Beneath The Surface is 0

In [79]:
token_lyrics = [tokenize(d) for d in lyrics]

# removing stopwords and stemming
porter = nltk.PorterStemmer()
lyrics_final = []
for song in token_lyrics:
    lyrics_stopwords = [word for word in song if word not in stopwords.words('english')]
    lyrics_stemming = [porter.stem(word) for word in lyrics_stopwords]
    lyrics_final.append(lyrics_stemming)
    

In [85]:
# join the letters together again to be able to use sklearn
lyrics_processed = [' '.join(song) for song in lyrics_final]

In [87]:
tfidf = TfidfVectorizer(norm='l2', min_df=0, use_idf=True, smooth_idf=False, sublinear_tf=True, tokenizer=tokenize)

representation = tfidf.fit_transform(lyrics_processed)

In [88]:
songs_name = ['Through Her Eyes', 'The Spirit Carries On', "Beneath The Surface", "Pull Me Under", "Wither", 
              "Another Brick In The Wall (Part II)", "Hey You", "Comfortably Numb", 'Time', "Wish You Were Here" ]
for i in range(10):
    for j in range(i+1,10):
        similarity = cosine_similarity(representation.toarray()[i], representation.toarray()[j])
        print("The similarity between the song {} and the song {} is {}".format(songs_name[i], songs_name[j], similarity))

The similarity between the song Through Her Eyes and the song The Spirit Carries On is 0.1036096976275753
The similarity between the song Through Her Eyes and the song Beneath The Surface is 0.06298287212867386
The similarity between the song Through Her Eyes and the song Pull Me Under is 0.11617477938607257
The similarity between the song Through Her Eyes and the song Wither is 0.0929824469158438
The similarity between the song Through Her Eyes and the song Another Brick In The Wall (Part II) is 0.014254177995545778
The similarity between the song Through Her Eyes and the song Hey You is 0.05272962806002318
The similarity between the song Through Her Eyes and the song Comfortably Numb is 0.11915146597694783
The similarity between the song Through Her Eyes and the song Time is 0.07696663086015136
The similarity between the song Through Her Eyes and the song Wish You Were Here is 0.01782579866900643
The similarity between the song The Spirit Carries On and the song Beneath The Surface i

the similarity between the songs seems to be lower for every pair of songs.    
I was thinking that stemming the words would give us a higher level of similarity, but i guess the stop word removal probably lowers even more the similarity.    
These operation dont affect the same way every lyrics and now the most similar songs are different than before.   
The most similar songs were: *The similarity between the song Through Her Eyes and the song Comfortably Numb is 0.19775882519287555* that now have a similarity score of 0.11915146597694783.      
on the other side the most similar ones now are *The similarity between the song Another Brick In The Wall (Part II) and the song Hey You is 0.14390777920909376* from a previous score of 0.18968897053542866

In [90]:
from nltk.corpus import treebank, stopwords
from nltk import ConditionalFreqDist 

# initialize a new conditional distribution
cfd = ConditionalFreqDist()

# get a list of English stopwords
stopwords_list = stopwords.words('english')

def is_noun(tag):
    return tag.lower() in ['nn','nns','nn$','nn-tl','nn+bez', 'nn+hvz', 
                           'nns$','np','np$','np+bez','nps', 'nps$','nr',
                           'np-tl','nrs','nr$']

for sentence in treebank.tagged_sents():
    for (index, tagtuple) in enumerate(sentence):
        (token, tag) = tagtuple
        token = token.lower()
        if token not in stopwords_list and is_noun(tag):
            window = sentence[index+1:index+5]
            for (window_token, window_tag) in window:
                window_token = window_token.lower()
                if window_token not in stopwords_list and is_noun(window_tag):
                    cfd[token][window_token] += 1

In [114]:
print(cfd['chairman'].max())

computers


In [97]:
print(cfd['profit'].max())

%


In [98]:
print(cfd['president'].max())

chief


In [99]:
print(cfd['investors'].max())

bonds


In [100]:
print(cfd['city'].max())

streets


In [101]:
print(cfd['chief'].max())

officer


'chief' seems to be the word people associate with a position of authority